packages/core/src/renderer/Renderer.ts

```diff
@@ -184,28 +116,190 @@ export const RenderStatic = async (
+  }
+  visited.add(name);
+  const shape = nameShapeMap[name];
+  if (shape.shapeType === "Group") {
+    const outSvg = await RenderGroup(
+      name,
+      graph,
+      shapeProps,
+      nameShapeMap,
+      visited,
+      interactiveProp
+    );
+    return outSvg;
+  } else {
+    const elem = await shapeMap[shape.shapeType]({
+      ...shapeProps,
+      shape,
+    });
+    if (!interactiveProp) {
+      return elem;
+    } else {
+      const g = document.createElementNS("http://www.w3.org/2000/svg", "g");
+      const { shapeType } = shape;
+      if (shapedefs[shapeType].isLinelike) {
+        g.setAttribute("pointer-events", "visibleStroke");
+      } else if (shapedefs[shapeType].isRectlike) {
+        g.setAttribute("pointer-events", "bounding-box");
+      } else {
+        g.setAttribute("pointer-events", "auto");
+      }
+      g.appendChild(elem);
+      const onMouseDown = (e: MouseEvent) => {
+        const { clientX, clientY } = e;
+        const { x: tempX, y: tempY } = getPosition(
+          { clientX, clientY },
+          interactiveProp.parentSVG
+        );
+        const {
+          width: bboxW,
+          height: bboxH,
+          x: bboxX,
+          y: bboxY,
+        } = (e.target as SVGSVGElement).getBBox({ stroke: true });
+        const minX = tempX - bboxX;
+        const maxX = shapeProps.canvasSize[0] - bboxW + (tempX - bboxX);
+        const minY = tempY - bboxY;
+        const maxY = shapeProps.canvasSize[1] - bboxH + (tempY - bboxY);
+
+        g.setAttribute("opacity", "0.5");
+        let dx = 0,
+          dy = 0;
+        const onMouseMove = (e: MouseEvent) => {
+          const { x, y } = getPosition(e, interactiveProp.parentSVG);
+          const constrainedX = clamp(x, minX, maxX);
+          const constrainedY = clamp(y, minY, maxY);
+          dx = constrainedX - tempX;
+          dy = tempY - constrainedY;
+          g.setAttribute(`transform`, `translate(${dx},${-dy})`);
+        };
+        const onMouseUp = () => {
+          g.setAttribute("opacity", "1");
+          document.removeEventListener("mouseup", onMouseUp);
+          document.removeEventListener("mousemove", onMouseMove);
+          interactiveProp.onDrag(
+            (shape.properties.name as StrV).contents,
+            dx,
+            dy
+          );
+        };
+        document.addEventListener("mouseup", onMouseUp);
+        document.addEventListener("mousemove", onMouseMove);
+      };
+      g.addEventListener("mousedown", onMouseDown);
+      return g;
+    }
+  }
+};
+
+const RenderGroupGraph = async (
+  graph: GroupGraph,
+  svg: SVGSVGElement,
+  shapeProps: {
+    labels: LabelCache;
+    canvasSize: [number, number];
+    variation: string;
+    pathResolver: PathResolver;
+  },
+  nameShapeMap: { [k: string]: Shape },
+  interactiveProp?: InteractiveProps
+) => {
+  const visited: Set<string> = new Set();
+  while (visited.size < graph.nodes().length) {
+    const rootName = findRoot(graph, visited);
```

Comment:
This feels backwards; why not instead keep track of which ones we have yet to visit, and then at each step of this loop, check if that set is empty and then pick one to do next? (Also, why are we running graph algorithms in the renderer at all? Does the compiler not return the right kind of structure for the render to just use directly?)

Reply:
The shape ordering outputted by the compiler does not respect the group ordering, only the layer ordering, since we run topsort only on the layer ordering.

Reply:
Of course, if we manage to get group ordering and layer ordering together into one ordering, then we can do what you said.